### PR TITLE
Fix BackupEntry operation annotation removal

### LINKED ISSUE #4540: BackupEntry reconcilition blocks for 24 hours or less when the gardenlet is restarted.
**How to categorize this issue?**

/area backup
/kind bug
/priority 3

**What happened**:
During normal Shoot reconciliation flow the `BackupEntry` in the Gardener cluster will be annotated with `gardener.cloud/operation=reconcile`. This will change the observed generation and the BE will be caught by the  [`backupEntryUpdate()`](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_controller.go#L49) function of the [BE controller](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry.go#L58).  The only condition that this function checks before putting the BE into the worker's queue is if the current generation matches the observed one. If so it puts the BE into the queue. Then into  [reconcileBackupEntry](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L97) the last observed generation will be [updated](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L108) with the current one. This will prevent status updates of BE to put the resource in the worker queue. Then the [Actuator](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L123) successfully reconcile the BE which status is updated on the next [line](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L139). After updating the BE status the `gardenlet` is suddenly restarted. We miss the last [step](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L144) where we remove the operation annotation.
On the next start the BE will be catch by the [backupEntryAdd()](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_controller.go#L28) which will send the BE into the queue only if:
1. [There is a deletion timestamp](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/controllerutils/miscellaneous.go#L100).
2. [Current generation is different from the observed one](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/controllerutils/miscellaneous.go#L104).
3. [The last operation is not succeeded](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/controllerutils/miscellaneous.go#L108).
4. [24h have passed since wast operation](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/controllerutils/miscellaneous.go#L116).

In this case, there is no deletion timestamp, the generations are the same, the last operation was successful and this 24h period is not expired.
So as the BE already has the `gardener.cloud/operation=reconcile` annotation, the next shoot reconciliation flow will not update the resource, which will not change the current generation, and thus the reconciliation will be skipped until 24h have passed since the last operation.
The `gardenlet` starts to complain that no controller is peaking up the BE and the cluster enters in an error state.

**What you expected to happen**:
The `gardener.cloud/operation=reconcile` to be removed.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
As a solution, we could either add to the worker queue BE with annotation gardener.cloud/operation=reconcile` when passing to [backupEntryAdd()](https://github.com/gardener/gardener/blob/077b71bddce44942018640f0bae69622a9103e66/pkg/gardenlet/controller/backupentry/backup_entry_controller.go#L28) or to remove it at the beginning of the same function.
**Environment**:

- Gardener version: v1.28.2
- Kubernetes version (use `kubectl version`): 1.19.10
- Cloud provider or hardware configuration: aws
- Others:

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area backup
/kind bug

**What this PR does / why we need it**:
Fixes a bug that caused the shoot reconciliation to fail if `gardenlet` was restarted just before removing the operation annotation from a `BackupEntry`.

**Which issue(s) this PR fixes**:
Fixes #4540

**Special notes for your reviewer**:
See the discussion in #4540 and in this PR regarding why this approach for fixing the issue was chosen over other possible approaches.

**Release note**:

```bugfix user
Fixed a bug that caused the shoot reconciliation to fail if `gardenlet` was restarted just before removing the operation annotation from a `BackupEntry`.
```

### CODE REVIEW

File: pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go [hunk 1]
```diff
@@ -110,6 +110,11 @@ func (r *reconciler) reconcileBackupEntry(ctx context.Context, gardenClient kube
 		operationType = gardencorev1beta1.LastOperationTypeRestore
 	}
 
+	if updateErr := removeGardenerOperationAnnotation(ctx, gardenClient.Client(), backupEntry); updateErr != nil {
+		backupEntryLogger.Errorf("Could not remove %q annotation: %+v", v1beta1constants.GardenerOperation, updateErr)
+		return reconcile.Result{}, updateErr
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: Let's say the object was annotated with the `restore` operation and the `gardenlet` exits right after those lines. When it comes up again, wouldn't it start performing a different operation because the information that it should actually be a "restore" is lost?

Comment (stoyanr) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: Yes, unfortunately it would. I guess the best way to deal with it is to update the status and remove the annotation in the same `Patch` operation. Is that ever possible, or do you have other suggestions how to handle it?

I also thought about the following sequence:
* Determine `operationType` based on the operation annotation.
* Update `status.lastOperation` (but not `status.observedGeneration`) yet.
* Remove the operation annotation.
* Update `status.observedGeneration`.

But I think this will enqueue the BackupEntry after the first status update. Still, perhaps it's better than what we have now.

Let me play a bit more with the above ideas.

Comment (rfranzke) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: >  I guess the best way to deal with it is to update the status and remove the annotation in the same Patch operation. Is that ever possible, or do you have other suggestions how to handle it?

No, unfortunately, this isn't possible in only one call.

Can't we update the `.status.lastOperation` to `<OPERATION> Progressing` first (including updating `.status.observedGeneration`) and remove the annotation afterwards? If the gardenlet exits after the `.status.lastOperation` was updated then it could still start to reconcile the resource by determining the operation type based on `.status.lastOperation.type`, or?

Comment (stoyanr) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: Removing the operation annotation after `status.observedGeneration` has been updated would lead to the original issue that I am trying to fix.

As discussed, I will look deeper into how this is handled for extension resources in the seed and perhaps try to introduce a similar handling here.

Comment (stoyanr) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: So basically, this is what happens for extension resources in the seed:
* Instead of checking the generation, resources are enqueued if they have the operation annotation, if the last operation was not successful, or if the deletion timestamp is not nil.
* The operation annotation is removed before the reconciliation if it's `reconcile`, and at the end of the reconciliation if it's `migrate` or `restore`.

This can lead to the following issues:
* If the extension is restarted after removing the operation annotation with value `reconcile`, but before updating the status, the reconciliation will be skipped. This is perhaps ok though since gardenlet will retry the flow at some point leading to another reconciliation.
* If the extension is restarted after updating the status, but before removing the operation annotation with value `restore`, or `migrate`, one redundant reconciliation will be enqueued. I guess this is also ok, since it's rare, and it would only happen for `restore` and `migrate` (the handling should be idempotent by design).

However, it cannot lead to either the original issue reported in #4540 (since the annotation itself, and not the generation, is what matters), or the issue you with my current fix (since the annotation is not removed if it's value is `restore` or `migrate`).

So basically what I would propose to do is to rework the enqueueing of `BackupEntry` to follow the above pattern, and even to get rid of checking the operation annotation in `strategy.go` (since the generation is no longer used to determine whether the resource should be enqueued).

Comment (rfranzke) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: Sounds good to me, thanks for looking into it!

Comment (stoyanr) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go]: OK, did as proposed above, please take another look.

**Comment by stoyanr** [comment]
/invite @plkokanov Could you please take a look regarding the handling of the `restore` operation? 

Meanwhile I tested the control plane migration, seems to work without any issues, still a second opinion would be great.

File: pkg/apis/core/v1beta1/helper/helper_test.go [hunk 2]
```diff
@@ -358,6 +358,16 @@ var _ = Describe("helper", func() {
...
 		)
 
+		DescribeTable("#HasOperationAnnotation",
+			func(objectMeta metav1.ObjectMeta, expected bool) {
+				Expect(HasOperationAnnotation(objectMeta)).To(Equal(expected))
+			},
+			Entry("reconcile", metav1.ObjectMeta{Annotations: map[string]string{v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationReconcile}}, true),
+			Entry("restore", metav1.ObjectMeta{Annotations: map[string]string{v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationRestore}}, true),
+			Entry("migrate", metav1.ObjectMeta{Annotations: map[string]string{v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationMigrate}}, true),
+			Entry("not present", metav1.ObjectMeta{}, false),
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/helper_test.go:369]: Nit: Add a test case with a value other than `reconcile`/`restore`/`migrate`

File: pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go [hunk 3]
```diff
@@ -77,6 +78,15 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		return reconcile.Result{}, err
 	}
 
+	// Remove the operation annotation if its value is not "restore"
```
Comment (rfranzke) [inline pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go:81]: Nit: Maybe extend the comment to explain why there is this special case for `restore`

File: pkg/apis/core/v1beta1/helper/helper_test.go (hunk 2 above)
Comment (stoyanr) [reply pkg/apis/core/v1beta1/helper/helper_test.go:369]: Done

File: pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go (hunk 3 above)
Comment (stoyanr) [reply pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go:81]: Done

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by plkokanov — APPROVED** [review]
/lgtm
